### PR TITLE
Add "Attach comment to issue" modal

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
1. Create button to open "Attach comment" modal in context menu.
2. Create modal to attach comment.
3. Create API to create issue comment & search issue.
4. Create notification post when comment is attached.

### CODE REVIEW

File: server/api.go [hunk 1]
```diff
@@ -625,53 +631,136 @@ func (p *Plugin) createIssue(c *UserContext, w http.ResponseWriter, r *http.Requ
 	p.writeAPIResponse(w, result)
 }
 
-func (p *Plugin) getPermaLink(postID string) string {
+func (p *Plugin) attachCommentToIssue(c *UserContext, w http.ResponseWriter, r *http.Request) {
+	var issue *gitlab.IssueRequest
+
+	if err := json.NewDecoder(r.Body).Decode(&issue); err != nil {
+		c.Log.WithError(err).Warnf("error decoding JSON body to attach comment to the issue")
```
Comment (manojmalik20) [inline server/api.go:638]:
```suggestion
		c.Log.WithError(err).Warnf("error decoding JSON body for attach comment to issue")
```

File: server/api.go [hunk 2]
```diff
@@ -625,53 +631,136 @@ func (p *Plugin) createIssue(c *UserContext, w http.ResponseWriter, r *http.Requ
...
+		p.writeAPIError(w, &APIErrorResponse{ID: "", Message: fmt.Sprintf("error decoding JSON body to attach comment to the issue. Error: %s", err.Error()), StatusCode: http.StatusBadRequest})
```
Comment (manojmalik20) [inline server/api.go:639]:
```suggestion
		p.writeAPIError(w, &APIErrorResponse{ID: "", Message: fmt.Sprintf("error decoding JSON body for attach comment to the issue. Error: %s", err.Error()), StatusCode: http.StatusBadRequest})
```
You can create a function which prints an error in server logs and also writes the error in the response and call that function here.

File: server/api.go [hunk 3]
```diff
@@ -625,53 +631,136 @@ func (p *Plugin) createIssue(c *UserContext, w http.ResponseWriter, r *http.Requ
...
+		return
+	}
+	if post == nil {
+		p.writeAPIError(w, &APIErrorResponse{ID: "", Message: fmt.Sprintf("failed to load post %s : not found", issue.PostID), StatusCode: http.StatusNotFound})
+		return
+	}
+
+	commentUsername, apiErr := p.getUsername(post.UserId)
+	if apiErr != nil {
+		p.writeAPIError(w, &APIErrorResponse{ID: "", Message: "failed to get username", StatusCode: http.StatusInternalServerError})
```
Comment (manojmalik20) [inline server/api.go:670]: Same here.

File: server/api.go [hunk 4]
```diff
@@ -625,53 +631,136 @@ func (p *Plugin) createIssue(c *UserContext, w http.ResponseWriter, r *http.Requ
...
+		// The original post was a reply
+		rootID = post.RootId
+	}
+
+	permalinkReplyMessage := fmt.Sprintf("[Message](%v) attached to GitLab issue [#%v](%v)", permalink, issue.IID, issue.WebURL)
+	reply := &model.Post{
+		Message:   permalinkReplyMessage,
+		ChannelId: post.ChannelId,
+		RootId:    rootID,
+		UserId:    c.UserID,
```
Comment (manojmalik20) [inline server/api.go:694]: Shouldn't this be the botID?

File: server/api.go [hunk 5]
```diff
@@ -625,53 +631,136 @@ func (p *Plugin) createIssue(c *UserContext, w http.ResponseWriter, r *http.Requ
...
+	}
+
+	_, appErr = p.API.CreatePost(reply)
+	if appErr != nil {
+		p.writeAPIError(w, &APIErrorResponse{ID: "", Message: fmt.Sprintf("failed to create notification post %s", issue.PostID), StatusCode: http.StatusInternalServerError})
```
Comment (manojmalik20) [inline server/api.go:699]: Use status code gotten in `appErr`.

File: server/api.go [hunk 6]
```diff
@@ -625,53 +631,136 @@ func (p *Plugin) createIssue(c *UserContext, w http.ResponseWriter, r *http.Requ
...
 	siteURL := *p.API.GetConfig().ServiceSettings.SiteURL
 
 	return fmt.Sprintf("%v/_redirect/pl/%v", siteURL, postID)
 }
 
+func (p *Plugin) searchIssues(c *UserContext, w http.ResponseWriter, r *http.Request) {
+	search := r.FormValue(search)
+	result, err := p.GitlabClient.SearchIssues(c.Ctx, c.GitlabInfo, search)
+	if err != nil {
+		c.Log.WithError(err).Warnf("unable to search issues in GitLab API")
```
Comment (manojmalik20) [inline server/api.go:715]:
```suggestion
		c.Log.WithError(err).Warnf("unable to search issues in GitLab")
```

File: server/api.go [hunk 7]
```diff
@@ -625,53 +631,136 @@ func (p *Plugin) createIssue(c *UserContext, w http.ResponseWriter, r *http.Requ
...
+		p.writeAPIError(w, &APIErrorResponse{ID: "", Message: fmt.Sprintf("unable to search issues in GitLab API. Error: %d", err), StatusCode: http.StatusInternalServerError})
```
Comment (manojmalik20) [inline server/api.go:716]:
```suggestion
		p.writeAPIError(w, &APIErrorResponse{ID: "", Message: fmt.Sprintf("unable to search issues in GitLab. Error: %s", err.Error()), StatusCode: http.StatusInternalServerError})
```
Why used `%d` here?

File: server/gitlab/api.go [hunk 8]
```diff
@@ -501,6 +504,54 @@ func (g *gitlab) CreateIssue(ctx context.Context, user *UserInfo, issue *IssueRe
...
+		&internGitlab.CreateIssueNoteOptions{
+			Body: &issueComment,
+		},
+		internGitlab.WithContext(ctx),
+	)
+	if respErr := checkResponse(resp); respErr != nil {
+		return nil, respErr
+	}
+	if err != nil {
+		return nil, errors.Wrap(err, "can't create issue comment in GitLab api")
```
Comment (manojmalik20) [inline server/gitlab/api.go:531]:
```suggestion
		return nil, errors.Wrap(err, "can't create issue comment in GitLab")
```
Fix everywhere.

File: server/plugin.go [hunk 9]
```diff
@@ -595,3 +595,24 @@ func (p *Plugin) HasGroupHook(ctx context.Context, user *gitlab.UserInfo, namesp
 
 	return found, err
 }
+
+// getUsername returns the GitLab username for a given Mattermost user,
+// if the user is connected to GitLab via this plugin.
+// Otherwise it return the Mattermost username. It will be escaped via backticks.
```
Comment (manojmalik20) [inline server/plugin.go:601]:
```suggestion
// Otherwise it returns the Mattermost username. It will be escaped via backticks.
```
What do you mean by "It will be escaped via backticks"?

File: webapp/src/components/attach_comment_to_issue_menu/attach_comment_to_issue.tsx [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        open: (postId: string) => {
+            type: string;
+            data: {
+                postId: string;
+            };
+        };
+    };
+};
+
+export default class AttachCommentToIssuePostMenuAction extends PureComponent<PropTypes> {
```
Comment (manojmalik20) [inline webapp/src/components/attach_comment_to_issue_menu/attach_comment_to_issue.tsx:22]: Convert to functional component.

File: webapp/src/components/attach_comment_to_issue_menu/index.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+interface CurrentState extends GlobalState {
+    plugin: pluginMethods;
+}
+
+interface OwnProps {
+    postId: string;
+}
+
+type Actions = {
```
Comment (manojmalik20) [inline webapp/src/components/attach_comment_to_issue_menu/index.ts:28]: Create this in the component file and import this here.

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+    };
+
+    debouncedSearchIssues = debounce(this.searchIssues, searchDebounceDelay);
+
+    onChange = (newValue: SingleValue<IssueSelection>) => {
+        const value = newValue?.value ?? null;
+        this.props.onChange(value);
+    }
+
+    isValid = () => {
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:95]: Same here.

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        }
+
+        const valid = Boolean(this.props.value);
+        this.setState({invalid: !valid});
+        return valid;
+    };
+
+    render() {
+        let issueError = null;
+        if (this.props.error) {
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:107]: Use `const` and ternary operators here.

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+            issueError = (
+                <p className='help-text error-text'>
+                    <span>{this.props.error}</span>
+                </p>
+            );
+        }
+
+        let serverError;
+        if (this.state.error) {
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:116]: Same here.

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+                        title='Warning Icon'
+                    />
+                    <span>{this.state.error}</span>
+                </p>
+            );
+        }
+
+        const requiredMsg = 'This field is required.';
+        let validationError = null;
+        if (this.props.required && this.state.invalid) {
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:130]: Same here.

File: webapp/src/components/input.tsx [hunk 16]
```diff
@@ -29,7 +29,9 @@ export default class Input extends PureComponent<PropTypes, StateTypes> {
     }
 
     handleChange = (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => {
-        this.props.onChange(e.target.value);
+        if (this.props.onChange) {
+            this.props.onChange(e.target.value);
+        }
```
Comment (manojmalik20) [inline webapp/src/components/input.tsx:34]:
```suggestion
        this.props.onChange?.(e.target.value);
```

File: webapp/src/components/modals/attach_comment_to_issue/attach_comment_to_issue.tsx [hunk 17]
```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {PureComponent} from 'react';
+import {Modal} from 'react-bootstrap';
+import {Theme} from 'mattermost-redux/types/preferences';
+import {Post} from 'mattermost-redux/types/posts';
+
+import FormButton from '../../form_button';
```
Comment (manojmalik20) [inline webapp/src/components/modals/attach_comment_to_issue/attach_comment_to_issue.tsx:9]: Absolute paths.

File: webapp/src/components/modals/attach_comment_to_issue/attach_comment_to_issue.tsx [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+import Input from '../../input';
+import Validator from '../../validator';
+
+import GitlabIssueSelector from '../../gitlab_issue_selector';
+import {getErrorMessage} from '../../../utils/user_utils';
+
+const initialState = {
```
Comment (manojmalik20) [inline webapp/src/components/modals/attach_comment_to_issue/attach_comment_to_issue.tsx:16]: Define it after defining the type.

File: webapp/src/components/modals/attach_comment_to_issue/attach_comment_to_issue.tsx [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+        if (created.error) {
+            const errMessage = getErrorMessage(created.error.message);
+            this.setState({error: errMessage, submitting: false});
+            return;
+        }
+
+        this.handleClose();
+    };
+
+    handleClose = () => {
```
Comment (manojmalik20) [inline webapp/src/components/modals/attach_comment_to_issue/attach_comment_to_issue.tsx:78]: Try to use `useCallback` or `useMemo` in every React component for optimization.

File: webapp/src/components/modals/attach_comment_to_issue/attach_comment_to_issue.tsx [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+    }
+}
+
+interface StateTypes {
+    submitting: boolean,
+    issueValue: Issue | null,
+    error: string,
+}
+
+export default class AttachCommentToIssueModal extends PureComponent<PropTypes, StateTypes> {
```
Comment (manojmalik20) [inline webapp/src/components/modals/attach_comment_to_issue/attach_comment_to_issue.tsx:42]: Convert to functional component.

File: webapp/src/components/modals/attach_comment_to_issue/index.ts [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+import {getPost} from 'mattermost-redux/selectors/entities/posts';
+import {ActionFunc, GenericAction} from 'mattermost-redux/types/actions';
+import {GlobalState} from 'mattermost-redux/types/store';
+
+import {id as pluginId} from '../../../manifest';
+import {closeAttachCommentToIssueModal, attachCommentToIssue} from '../../../actions';
+
+import AttachCommentToIssueModal from './attach_comment_to_issue';
+
+interface pluginMethods {
```
Comment (manojmalik20) [inline webapp/src/components/modals/attach_comment_to_issue/index.ts:15]: These are not methods.

File: webapp/src/components/modals/attach_comment_to_issue/index.ts [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+    close: () => {
+        type: string;
+    };
+    create: (payload: CommentBody) => Promise<{
+        error?: ErrorType;
+    }>;
+};
+
+const mapStateToProps = (state: CurrentState) => {
+    const postId = state[`plugins-${pluginId}` as plugin].attachCommentToIssueModalForPostId;
```
Comment (manojmalik20) [inline webapp/src/components/modals/attach_comment_to_issue/index.ts:34]:
```suggestion
    const postId = state[`plugins-${pluginId}` as plugin].postIdForAttachCommentToIssueModal;
```

File: webapp/src/images/icons/gitlab.tsx [hunk 23]
```diff
@@ -1,30 +1,30 @@
...
 
 interface PropTypes {
-    type: String;
-}
+    type: string;
+};
 
-const GitLabIcon = ({type}: PropTypes) => {
-    const iconStyle = type === 'menu' ? {flex: '0 0 auto', width: '20px', height: '20px', borderRadius: '50px', padding: '2px'} : {};
+export default class GitLabIcon extends PureComponent<PropTypes> {
```
Comment (manojmalik20) [inline webapp/src/images/icons/gitlab.tsx:10]: Convert to functional.

**Review by manojmalik20 — CHANGES_REQUESTED** [review]

File: server/api.go (hunk 7 above)
Comment (raghavaggarwal2308) [reply server/api.go:716]: Typing mistake!

File: server/api.go (hunk 2 above)
Comment (raghavaggarwal2308) [reply server/api.go:639]: discussed

File: webapp/src/components/attach_comment_to_issue_menu/index.tsx [hunk 24]
```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {MouseEvent, useCallback} from 'react';
+import PropTypes from 'prop-types';
+import { useDispatch, useSelector } from 'react-redux';
+import { GlobalState } from 'mattermost-redux/types/store';
```
Comment (manojmalik20) [inline webapp/src/components/attach_comment_to_issue_menu/index.tsx:7]:
```suggestion
import {useDispatch, useSelector} from 'react-redux';
import {GlobalState} from 'mattermost-redux/types/store';
```
Fix these errors everywhere.

File: webapp/src/components/attach_comment_to_issue_menu/index.tsx [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+interface CurrentState extends GlobalState {
+    plugin: states;
+}
+
+const AttachCommentToIssuePostMenuAction = ({postId}: PropTypes) => {
+    const {show} = useSelector((state: CurrentState) => {
+        const post = getPost(state, postId);
+        const systemMessage = post ? isSystemMessage(post) : true;
```
Comment (manojmalik20) [inline webapp/src/components/attach_comment_to_issue_menu/index.tsx:30]: I think I suggested to change this in another PR. Do that change here and everywhere.

File: webapp/src/components/create_issue_menu/index.tsx [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+        const systemMessage = Boolean(!post || isSystemMessage(post));
+    
+        return {
+            show: state[`plugins-${pluginId}` as plugin].connected && !systemMessage,
+        };
+    })
+
+    const dispatch = useDispatch();
+
+    const handleClick = (e: MouseEvent<HTMLButtonElement> | Event) => {
```
Comment (manojmalik20) [inline webapp/src/components/create_issue_menu/index.tsx:38]: Use `useCallback` here.

File: webapp/src/components/gitlab_assignee_selector/index.tsx [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+    projectName: string;
+    theme: Theme;
+    selectedAssignees: SelectionType[];
+    onChange: (assignees: OnChangeType) => void;
+};
+
+const GitlabAssigneeSelector = ({projectID, projectName, theme, selectedAssignees, onChange}: PropTypes) => {
+    const dispatch = useDispatch();
+
+    const loadAssignees = async () => {
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_assignee_selector/index.tsx:22]: Use `useCallback` everywhere.

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 28]
```diff
@@ -25,43 +25,42 @@ interface PropTypes {
...
+    const isValid = useCallback(() => {                
+        if (!props.required) {
+            return true;
         }
-    }
+        
+        const valid = Boolean(props.value);
+        setInvalid(!valid);
+        return valid;
+    }, [props.value])
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:40]: Shouldn't this also contain `props.required`? Test this and find out.

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 29]
```diff
@@ -25,43 +25,42 @@ interface PropTypes {
...
+                props.removeValidate(props.name);
+            }
         }
-    }
-
-    componentDidUpdate() {
-        if (this.state.invalid) {
-            this.isValid();
+    }, [])
+
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:49]:
```suggestion

```

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 30]
```diff
@@ -25,43 +25,42 @@ interface PropTypes {
...
-    }
-
-    componentDidUpdate() {
-        if (this.state.invalid) {
-            this.isValid();
+    }, [])
+    
+    useEffect(() => {
+        if (props.addValidate && props.name) {            
+            props.addValidate(props.name, isValid);
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:52]: This code for `componentDidMount` can exist in the same `useEffect` that you used above for `componentWillUnmount`.

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 31]
```diff
@@ -25,43 +25,42 @@ interface PropTypes {
...
         }
-    }
+        if (invalid) {            
+            isValid();
+        }
+    }, [props.value])
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:57]: Why is this here? Isn't this useEffect for `componentDidMount`?

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 32]
```diff
@@ -73,95 +72,77 @@ export default class GitlabIssueSelector extends PureComponent<PropTypes, StateT
             return issues.map((issue) => {
                 const projectParts = issue.web_url.split('/');
                 let prefix = '';
+                //Extract "username/projectName" from the issueURL parts
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:75]:
```suggestion
                // Extract "username/projectName" from the issueURL parts
```
I think the comment can be more descriptive or the code can be more readable. We can discuss this.

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 33]
```diff
@@ -25,43 +25,42 @@ interface PropTypes {
...
+    }, [props.value])
 
-    handleIssueSearchTermChange = (inputValue: string) => {
-        return this.debouncedSearchIssues(inputValue);
+    const handleIssueSearchTermChange = (inputValue: string) => {
+        return debouncedSearchIssues(inputValue);
     };
 
-    searchIssues = async (text: string) => {
+    const searchIssues = async (text: string) => {
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:63]: Try to use `useCallback` or `useMemo` here.

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 34]
```diff
@@ -73,95 +72,77 @@ export default class GitlabIssueSelector extends PureComponent<PropTypes, StateT
...
-                    {issueError}
-                    <div className={'help-text'}>
-                        {'Returns issues sorted by most recently created.'} <br/>
-                    </div>
-                </>
-            </Setting>
-        );
-    }
+        props.onChange(value);
+    }, [])
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:93]: Shouldn't the array contain `props.onChange`? Test this.

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 35]
```diff
@@ -73,95 +72,77 @@ export default class GitlabIssueSelector extends PureComponent<PropTypes, StateT
...
+
+    const issueError = (props.error) ? (
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:95]:
```suggestion
    const issueError = props.error ? (
```

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 36]
```diff
@@ -73,95 +72,77 @@ export default class GitlabIssueSelector extends PureComponent<PropTypes, StateT
...
+        <p className='help-text error-text'>
+            <span>{props.error}</span>
+        </p>
+    ) : null;
+
+    const serverError = (error) ? (
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:101]:
```suggestion
    const serverError = error ? (
```

File: webapp/src/components/gitlab_issue_selector.tsx [hunk 37]
```diff
@@ -73,95 +72,77 @@ export default class GitlabIssueSelector extends PureComponent<PropTypes, StateT
...
+            <i
+                className='fa fa-warning'
+                title='Warning Icon'
+            />
+            <span>{error}</span>
+        </p>
+    ) : null;
+
+    const requiredMsg = 'This field is required.';
+    const validationError = (props.required && invalid) ? (
```
Comment (manojmalik20) [inline webapp/src/components/gitlab_issue_selector.tsx:112]:
```suggestion
    const validationError = props.required && invalid ? (
```

File: webapp/src/components/input.tsx [hunk 38]
```diff
@@ -18,79 +18,67 @@ interface PropTypes {
...
-        super(props);
-        this.state = {invalid: false};
-    }
-
-    handleChange = (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => {
-        if (this.props.onChange) {
-            this.props.onChange(e.target.value);
-        }
+const Input = (props: PropTypes) => {   
+    const handleChange = (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => {
```
Comment (manojmalik20) [inline webapp/src/components/input.tsx:22]: Try to use `useCallback` here.

File: webapp/src/components/modals/attach_comment_to_issue/index.tsx [hunk 39]
```diff
@@ -0,0 +1,173 @@
...
+import {id as pluginId} from 'src/manifest';
+import GitlabIssueSelector from 'src/components/gitlab_issue_selector';
+import {getErrorMessage} from 'src/utils/user_utils';
+import {closeAttachCommentToIssueModal, attachCommentToIssue} from 'src/actions';
+
+interface PropTypes {  
+    theme: Theme,
+}
+
+interface states {
```
Comment (manojmalik20) [inline webapp/src/components/modals/attach_comment_to_issue/index.tsx:23]:
```suggestion
interface State {
```

File: webapp/src/images/icons/gitlab.tsx [hunk 40]
```diff
@@ -1,30 +1,30 @@
...
-                    height='14'
-                    style={iconStyle}
-                >
-                    <path d='M21 14l-9 7l-9 -7l3 -11l3 7h6l3 -7z'/>
-                </svg>
-            </span>
-        );
-    }
+const GitLabIcon = ({type}: PropTypes) => {
+    const iconStyle = (type === 'menu') ? {flex: '0 0 auto', width: '20px', height: '20px', borderRadius: '50px', padding: '2px'} : {};
```
Comment (manojmalik20) [inline webapp/src/images/icons/gitlab.tsx:11]:
```suggestion
    const iconStyle = type === 'menu' ? {flex: '0 0 auto', width: '20px', height: '20px', borderRadius: '50px', padding: '2px'} : {};
```

File: webapp/src/components/gitlab_issue_selector.tsx (hunk 32 above)
Comment (raghavaggarwal2308) [reply webapp/src/components/gitlab_issue_selector.tsx:75]: discussed

File: webapp/src/components/gitlab_issue_selector.tsx (hunk 30 above)
Comment (raghavaggarwal2308) [reply webapp/src/components/gitlab_issue_selector.tsx:52]: discussed

File: webapp/src/components/issue_attribute_selector.tsx [hunk 41]
```diff
@@ -18,114 +19,101 @@ interface PropTypes {
...
-        );
-    }
+        props.onChange(null);
+    }, [props.selection, props.isMulti, props.onChange])
+
+    const onChangeHandler =  useCallback((newValue: OnChangeValue<OnChangeType, boolean>) => {
+        props.onChange(newValue as OnChangeType)
+    }, [props.onChange]);
+
+    const noOptionsMessage = props.projectName ? 'No options' : 'Please select a project first';
```
Comment (manojmalik20) [inline webapp/src/components/issue_attribute_selector.tsx:85]: You can use `useMemo` for this.

File: webapp/src/components/modals/attach_comment_to_issue/index.tsx [hunk 42]
```diff
@@ -15,27 +14,19 @@ import {id as pluginId} from 'src/manifest';
...
-    postIdForAttachCommentToIssueModal: string;
-    attachCommentToIssueModalVisible: boolean
-}
-
-interface CurrentState extends GlobalState {
-    plugin: states;
-}
-
 const AttachCommentToIssueModal = ({theme}: PropTypes) => {
     const [validator, setValidator] = useState(new Validator())
```
Comment (manojmalik20) [inline webapp/src/components/modals/attach_comment_to_issue/index.tsx:24]: Try using `useMemo` instead of `useState` here.

File: webapp/src/components/modals/create_issue/index.tsx [hunk 43]
```diff
@@ -0,0 +1,284 @@
...
+        return {
+            visible: state[`plugins-${pluginId}` as plugin].isCreateIssueModalVisible,
+            post,
+            title,
+            channelId,
+        };
+    });
+
+    const dispatch = useDispatch();
+
```
Comment (manojmalik20) [inline webapp/src/components/modals/create_issue/index.tsx:52]:
```suggestion
```

File: webapp/src/components/modals/create_issue/index.tsx [hunk 44]
```diff
@@ -0,0 +1,284 @@
...
+import {GlobalState} from 'src/types/global_state';
+import {usePrevious} from 'src/hooks/use_previous';
+
+const MAX_TITLE_LENGTH = 256;
+
+type PropTypes = {
+    theme: Theme;
+};
+
+const CreateIssueModal = (props: PropTypes) => {
```
Comment (manojmalik20) [inline webapp/src/components/modals/create_issue/index.tsx:26]: Extract fields using destructuring.

File: webapp/src/components/modals/create_issue/index.tsx [hunk 45]
```diff
@@ -0,0 +1,284 @@
...
+    const prevChannelId = usePrevious(channelId);
+    const prevTitle = usePrevious(title);
+
+    useEffect(() => {        
+        if (post && !prevPost) {
+            setIssueDescription(post.message);
+        } else if (channelId && (channelId !== prevChannelId || title !== prevTitle)) {
+            setIssueTitle(title.substring(0, MAX_TITLE_LENGTH));
+        }
+    }, [props, prevPost, prevChannelId, prevTitle, channelId, title, post]);
```
Comment (manojmalik20) [inline webapp/src/components/modals/create_issue/index.tsx:63]:
```suggestion
    }, [channelId, title, post]);
```
Is adding the `prev` fields really needed here? Their value changes every time the value of the original field changes.

And what's the point of mentioning the `props` here. Those only contain the theme. Should we run the function whenever the theme changes?

File: webapp/src/components/modals/create_issue/index.tsx [hunk 46]
```diff
@@ -0,0 +1,284 @@
...
+
+    // handle issue creation after form is populated
+    const handleCreate = useCallback(async (e: React.FormEvent<HTMLFormElement> | Event) => {
+        e.preventDefault();
+
+        if (!validator.validate() || !issueTitle) {
+            console.log(issueTitle);
+
```
Comment (manojmalik20) [inline webapp/src/components/modals/create_issue/index.tsx:71]:
```suggestion
```

File: webapp/src/components/modals/create_issue/index.tsx [hunk 47]
```diff
@@ -0,0 +1,284 @@
...
+        if (created.error) {
+            const errMessage = getErrorMessage((created as {error: ErrorType}).error.message);
+            setError(errMessage);
+            setShowErrors(true);
+            setSubmitting(false);
+            return;
+        }
+
+        handleClose();
+    }, [props, issueTitle, issueTitleValid, issueDescription, validator, labels, milestone, assignees, project, channelId]);
```
Comment (manojmalik20) [inline webapp/src/components/modals/create_issue/index.tsx:102]: Is there really a need for props here?

File: webapp/src/components/modals/create_issue/index.tsx [hunk 48]
```diff
@@ -0,0 +1,284 @@
...
+    const handleIssueTitleChange = useCallback((issueTitle: string) => {        
+        setIssueTitle(issueTitle);
+        if (issueTitle && !issueTitleValid) {
+            setIssueTitleValid(true);
+        }
+    }, [issueTitleValid]);
+
+    const handleIssueDescriptionChange = useCallback((issueDescription: string) => setIssueDescription(issueDescription), []);
+
+    const renderIssueAttributeSelectors = useCallback(() => {
```
Comment (manojmalik20) [inline webapp/src/components/modals/create_issue/index.tsx:135]: Can you try to use `useMemo` here?

File: webapp/src/types/common.d.ts [hunk 49]
```diff
@@ -9,4 +9,9 @@ type ErrorType = {
     message: string;
 }
 
-type plugin = "plugin"
+type plugin = 'plugins-com.github.manland.mattermost-plugin-gitlab'
+
+type GetOptions = (projectID?: number) => (dispatch: any) => Promise<{
```
Comment (manojmalik20) [inline webapp/src/types/common.d.ts:14]: You can import a type for `dispatch` from the `mattermost-redux`.

File: webapp/src/components/issue_attribute_selector.tsx [hunk 50]
```diff
@@ -54,53 +54,55 @@ const IssueAttributeSelector = (props: PropTypes) => {
...
+        onChange(newValue as OnChangeType)
+    }, [onChange]);
 
-    const noOptionsMessage = props.projectName ? 'No options' : 'Please select a project first';
+    const noOptionsMessage = useMemo(() => projectName ? 'No options' : 'Please select a project first', [projectName]);
 
     return (
-        <Setting {...props}>
+        <Setting
+            label={label}
```
Comment (manojmalik20) [inline webapp/src/components/issue_attribute_selector.tsx:89]: Are you sure it doesn't need any other prop?

File: webapp/src/components/modals/create_issue/index.tsx [hunk 51]
```diff
@@ -49,26 +49,23 @@ const CreateIssueModal = (props: PropTypes) => {
...
 
-    useEffect(() => {        
+    useEffect(() => {
         if (post && !prevPost) {
             setIssueDescription(post.message);
         } else if (channelId && (channelId !== prevChannelId || title !== prevTitle)) {
             setIssueTitle(title.substring(0, MAX_TITLE_LENGTH));
         }
-    }, [props, prevPost, prevChannelId, prevTitle, channelId, title, post]);
+    }, [channelId, title, post]);
```
Comment (manojmalik20) [inline webapp/src/components/modals/create_issue/index.tsx:62]: Did you test it after removing the "prev" fields from here?

File: webapp/src/components/modals/create_issue/index.tsx [hunk 52]
```diff
@@ -132,15 +129,15 @@ const CreateIssueModal = (props: PropTypes) => {
 
     const handleIssueDescriptionChange = useCallback((issueDescription: string) => setIssueDescription(issueDescription), []);
 
-    const renderIssueAttributeSelectors = useCallback(() => {        
+    const renderIssueAttributeSelectors = useMemo(() => {
```
Comment (manojmalik20) [inline webapp/src/components/modals/create_issue/index.tsx:132]:
```suggestion
    const issueAttributeSelectors = useMemo(() => {        
```

**Review by manojmalik20 — APPROVED** [review]
